signer/src/transaction_coordinator.rs

```diff
@@ -321,6 +330,48 @@ where
         Ok(())
     }
 
+    /// Submit the rotate key tx for the latest DKG shares, if the aggregate key
+    /// differs from the one in the smart contract registry
+    #[tracing::instrument(skip_all, fields(bitcoin_chain_tip))]
+    async fn check_and_submit_rotate_key_transaction(
+        &mut self,
+        bitcoin_chain_tip: &model::BitcoinBlockHash,
+        aggregate_key: &PublicKey,
+    ) -> Result<(), Error> {
+        let last_dkg = self
+            .context
+            .get_storage()
+            .get_latest_encrypted_dkg_shares()
+            .await?;
+
+        // If we don't have DKG shares nothing to do here
+        let Some(last_dkg) = last_dkg else {
+            return Ok(());
+        };
+
+        let current_aggregate_key = self
+            .context
+            .get_stacks_client()
+            .get_current_signers_aggregate_key(&self.context.config().signer.deployer)
+            .await?;
+
+        // If the latest DKG aggregate key matches on-chain data, nothing to do here
+        if Some(last_dkg.aggregate_key) == current_aggregate_key {
+            return Ok(());
+        }
+
+        let wallet = self.get_signer_wallet(bitcoin_chain_tip).await?;
+
+        self.construct_and_sign_rotate_key_transaction(
+            bitcoin_chain_tip,
+            aggregate_key,
```

Comment:
Right, and according to the docs of `StacksTransactionSignRequest` (where this end up) `It is used to identify the signing set for the transaction.`, so even if we just did a DKG with some new signers we want to use the "old" signing set to do the handoff. (I'll change it, with a fallback to `aggregate_key` if we have no `current_aggregate_key`).

Something I find confusing though is that, while we do validation right checking for dkg shares associated with that aggregate key, then when signing we use `SignerWallet::load`, that appears to load always the last key rotation. I guess this is going to work (probably almost) always, but it's based on some potentially non obvious assumptions.

Reply:
Yeah, doing rotate keys correctly is not obvious, and I think this is partly because we use confusing terminology. Also, things are structured confusingly too. `SignerWallet::load` should probably be renamed or something too. Yeah, it's not the best.